in-progress/0003-native-merkle-trees.md

```diff
@@ -0,0 +1,212 @@
+|                      |                                                                                               |
+| -------------------- | --------------------------------------------------------------------------------------------- |
+| Owners               | @PhilWindle                                                                                   |
+| Approvers            | @just-mitch @alexghr                                                                          |
+| Target Approval Date | 2024-06-21                                                                                    |
+
+
+## Executive Summary
+
+This design attempts to solve the problem of slow sync and merkle tree insertion performance.
+
+
+## Introduction
+
+We require high performance merkle tree implementations both to ensure nodes can stay synched to the network and sequencers/provers can advance the state as required to build blocks. Our cuirrent TS implementations are limited in their single-threaded nature and the unavoidable constraint of have to repeatedly call into WASM to perform a hash operation.
+
+Some analysis of the quantity of hashing and the time required can be found [here](https://hackmd.io/@aztec-network/HyfTK9U5a?type=view).
+
+This design proposes the creation of a set of multi-threaded merkle tree implementations in C++ using LMDB. It builds upon some previous prototyping to develop concurrent indexed tree insertions.
+
+## Implementation
+
+There are many parts to this design, we will walk through them individiually and discuss the choices made at each stage.
+
+### Overall Architecture
+
+A new C++ binary, World State, will be created that will be started by the node software. It will be configured with the location in which Merkle Tree data should be stored. It will then accept and respond with msgpack-ed messages over one or more streams. The initial implementation will simply used stdio, but this will be absrtacted such that this could be replaced by other stream-based mechanisms.
```

Comment:
Why create it as a separate binary instead of a [nodejs cpp add-on](https://nodejs.org/api/addons.html), so we don't need to manage a separate process? I'd also expect the base cost of calling into an add-on to be cheaper than calling into another process.

Reply:
This could maybe work too